docs/content/config/security/mail_crypt.md

```diff
@@ -64,3 +70,71 @@ Official Dovecot documentation: https://doc.dovecot.org/configuration_manual/mai
     ```
 
 This should be the minimum required for encryption of the mail while in storage.
+
+
+## Encrypted User Keys
+
+1. Create `10-custom.conf` and populate it with the following:
+
+    ```
+    mail_attribute_dict = file:%h/Maildir/dovecot-attributes
+    mail_plugins = $mail_plugins mail_crypt
+    mail_debug= yes
+
+    plugin {
+      mail_crypt_curve = secp521r1
+      mail_crypt_save_version = 2
+      mail_crypt_require_encrypted_user_key = yes
+    }
+    ```
+
+2. Create `auth-passwdfile.inc` and populate it with the following:
+
+    ```
+    # Authentication for passwd-file users. Included from 10-auth.conf.
+    #
+    # passwd-like file with specified location.
+    # <doc/wiki/AuthDatabase.PasswdFile.txt>
+
+    passdb {
+      driver = passwd-file
+      args = scheme=CRYPT username_format=%u /etc/dovecot/userdb
+      override_fields = userdb_mail_crypt_private_password=%w userdb_mail_crypt_save_version=2
```

Comment:
have you tried generating sha256 for the private key password?

Reply:
I did not, no. Was that the alt commented line you had in your comment on the other issue?

Reply:
I see, nevermind. Let me give it a shot

Reply:
```
Error: Mailbox INBOX: UID=9: read() failed: read(/var/mail/xx/xx/cur/1625932501.M833286P2172.ip-xxx.internal,S=7230,W=7371:2,S) failed: Private key not available: Cannot decrypt key 39f8ca52927fff3d5be10a7e52e06ae0e40e85845f89851b076c39c014ed8b21: error:03070068:bignum routines:BN_mpi2bn:encoding error (FETCH BODY[])
```

Reply:
Should `%{sha256:password}` be changed somehow?

Reply:
Maybe it should be `%{sha256:w}`

I've also read that otherwise some passwords might crash this scheme (like a password with a percentage sign in it). And hashing makes them safer as well.

You could also ask users just to give their hash instead of the password, but that's maybe a different usecase.

Reply:
Hmm, still no luck. I'm using
```
  override_fields = userdb_mail_crypt_private_password=%{sha256:w} userdb_mail_crypt_save_version=2
```

but I see: `dovecot: auth: Error: passwd-file(xxxxx): Failed to expand override_fields: Unknown variable '%w'`

Reply:
I'm also using 2.3.14 and the docs say this is available in 2.2.27.

Reply:
Do I need to be doing anything special with doveadm when I create the user's key?

Reply:
I would expect that you also need to pass in the sha256 value of the pw during construction. But it looks like the variable expansion is failing. I got that syntax from an email in the dovecot mailinglist. Maybe there is a place where the variable expansions are documented?

Reply:
Ok, found it: https://doc.dovecot.org/configuration_manual/config_file/config_variables/

It should be as you said. `%{sha256:password}` but then for the private key lock you should also do the Sha256 of the password instead of the raw password

Reply:
That works!! Changes made to sha256sum the passwords before they're used on the keys! Thanks!